core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/zookeeper/install-upgrade.yml

```diff
@@ -0,0 +1,78 @@
+---
+
+- name: Download Zookeeper binaries
+  include_role:
+    name: download
+    tasks_from: download_file
+  vars:
+    file_name: "{{ zookeeper_defaults.zookeeper_bin_filename }}"
+
+- name: Create Zookeeper directories
+  become: yes
+  file:
+    path: "/opt/zookeeper-{{ zookeeper_defaults.zookeeper_version }}"
+    recurse: yes
+    owner: "{{ zookeeper_defaults.zookeeper_user }}"
+    group: "{{ zookeeper_defaults.zookeeper_group }}"
+    mode: u=rwx,g=rx,o=rx
+    state: directory
+
+- name: Uncompress Zookeeper installation file
+  unarchive:
+    remote_src: yes
+    src: "{{ download_directory }}/{{ zookeeper_defaults.zookeeper_bin_filename }}"
+    dest: /opt/zookeeper-{{ zookeeper_defaults.zookeeper_version }}
+    creates: "/opt/zookeeper-{{ zookeeper_defaults.zookeeper_version }}/bin"
+    extra_opts: [--strip-components=1]
+    mode: u=rwx,g=rx,o=rx
+    owner: "{{ zookeeper_defaults.zookeeper_user }}"
+    group: "{{ zookeeper_defaults.zookeeper_group }}"
+
+- name: Copy configuration from previous version
+  copy:
+    remote_src: yes
+    src: /opt/zookeeper/conf/
+    dest: /opt/zookeeper-{{ zookeeper_defaults.zookeeper_version }}/conf
+    mode: preserve
+
+- name: Link /opt/zookeeper to recently installed version
+  file:
+    dest: /opt/zookeeper
+    state: link
+    src: /opt/zookeeper-{{ zookeeper_defaults.zookeeper_version }}
+    force: yes
+
+- name: Remove previous version binaries
+  file:
+    path: /opt/zookeeper-{{ before_upgrade_zookeeper_version }}
+    state: absent
+
+- name: Improve Zookeeper service and run zookeeper service using symbolic link instead of versioned path
+  lineinfile:
+    path: /lib/systemd/system/zookeeper.service
+    state: present
+    regexp: '^ExecStart=/opt/zookeeper-.*'
+    line: "ExecStart=/opt/zookeeper/bin/zkServer.sh start-foreground"
+
+- name: Check if any snapshots exists in data dir what is necessary in order to run zookeeper after upgrade
+  find:
+    paths: "{{ zookeeper_defaults.zookeeper_data_dir }}/version-2"
+    patterns: 'snapshot*.'
+    use_regex: yes
+  register: snapshot_exists
+
+# From 3.5.5 version, ZooKeeper is not able to start when no snapshot files present, what is valid scenario in 3.4.X version. Empty snapshot downloaded from Zookeeper's Jira ticket. 
+- name: Copy empty snapshot if not exists
+  copy:
+    dest: "{{ zookeeper_defaults.zookeeper_data_dir }}/version-2" 
+    src: roles/zookeeper/files/snapshot.0
```

Comment:
Since it seems this file is used only for upgrade, maybe it's better to put it somewhere inside upgrade role?

Alternative approach (not tested) is to force Zookeeper to generate a snapshot file when needed:
1. Set `snapshot.trust.empty` property in config file or use `-Dzookeeper.snapshot.trust.empty=true` property when there is no any snapshot.
2. Start ZK to generate a snapshot.
3. Remove the property.
4. Restart ZK.

Based on: https://issues.apache.org/jira/browse/ZOOKEEPER-3056?focusedCommentId=16974905&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16974905

Reply:
I was already testing this scenario and it didn't solve the problem. ZK will run with the property but we don't want to keep the property forever. if the ZK doesn't generate snapshot after startup and we will remove the property and we will restart zk, the problem will still exist. We can't force to generate snapshot.